### PR TITLE
adding the indexOfRow method to RowTable

### PR DESCRIPTION
**Pull request by Doriko**
Needed for recovering an index of a row in a parent table.

### CODE REVIEW

**Comment by Doriko** [comment]
Adding indexOfRow to interface, stub, etc.

File: user/src/main/java/org/tessell/widgets/HasRows.java [hunk 1]
```diff
@@ -23,6 +24,12 @@
   /** Removes row {@code view}. */
   boolean removeRow(final IsWidget view);
 
+  /** Assumes {@code row} is a table row and returns its index */
+  int indexOfRow(final Widget row);
```
Comment (stephenh) [inline user/src/main/java/org/tessell/widgets/HasRows.java]: Given the interface is primarily used for testing (e.g. the DOM-decoupled IsWidget interface), I don't think we need this indexOfRow(Widget) overload in HasRows.

File: user/src/main/java/org/tessell/widgets/StubRowTable.java [hunk 2]
```diff
@@ -90,4 +92,14 @@ public void add(int index, IsWidget a) {
...
   }
+
+  @Override
+  public int indexOfRow(Widget row) {
+    return rows.indexOf(row);
+  }
+
+  @Override
+  public int indexOfRow(com.google.gwt.user.client.ui.IsWidget row) {
+    return rows.indexOf(row.asWidget());
```
Comment (stephenh) [inline user/src/main/java/org/tessell/widgets/StubRowTable.java]: In a unit test, row.asWidget() will blow up saying "i'm not a real widget". And rows is a collection of IsWidgets anyway, so I think just rows.indexOf(row) is what you want.

File: user/src/main/java/org/tessell/widgets/HasRows.java (hunk 1 above)
Comment (Doriko) [inline user/src/main/java/org/tessell/widgets/HasRows.java]: Oh, I see. Will move.

**Comment by Doriko** [comment]
Moved indexOfRow to another interface

File: user/src/main/java/org/tessell/widgets/IsRowTable.java [hunk 3]
```diff
@@ -3,8 +3,16 @@
...
 import org.tessell.util.ListDiff.ListLike;
 
+import com.google.gwt.user.client.ui.Widget;
+
 public interface IsRowTable extends IsWidget, HasRows {
 
   ListLike<IsWidget> getRowsPanel();
 
+  /** Assumes {@code row} is a table row and returns its index */
+  int indexOfRow(final Widget row);
```
Comment (stephenh) [inline user/src/main/java/org/tessell/widgets/IsRowTable.java]: Hehe...no... :-)

The other interface, HasRows, was fine.

I just meant you have two methods here:

* indexOfRow(Widget) -- is a real dom widget, can only be used at runtime
* indexOfRow(IsWidget) -- can be either a real dom widget (when IsWidget is implemented by the real widgets) or a test widget, so can be used at either runtime or test time.

Given indexOfRow(IsWidget) can be used in more places that indexOfRow, I believe it's sufficient to provide only the indexOfRow(IsWidget) overload, especially in the HasRows/IsRowTable interfaces (given they are explicitly meant for unit tests, for which indexOfRow(Widget) is not useful).

If you want to keep the two overloads of indexOfRow (one for IsWidget and one for Widget) on the concrete RowTable type (e.g. for a codebase that didn't case about unit testing, had only Widgets, and so wanted to call RowTable.indexOfRow(Widget) directly), that would be fine--just trying pointing out it (indexOfRow(Widet)) is useful in the concrete RowTable, but not really the test interfaces.

File: user/src/main/java/org/tessell/widgets/RowTable.java [hunk 4]
```diff
@@ -62,6 +62,18 @@ public void addRow(final IsWidget isWidget) {
...
+  /** Assumes {@code row} is a table row and returns its index */
+  @Override
+  public int indexOfRow(final Widget row) {
+    return rows.indexOf(row);
+  }
+
+  /** Assumes {@code row} is a table row and returns its index */
+  @Override
+  public int indexOfRow(final org.tessell.gwt.user.client.ui.IsWidget row) {
+    return rows.indexOf(row);
```
Comment (stephenh) [inline user/src/main/java/org/tessell/widgets/RowTable.java]: If you look at where "rows" is maintained, the contents (in this class) are always the real Widgets, so you should call row.asWidget() here.

**Comment by stephenh** [comment]
Pushed into S3 as 2.18.11. Thanks!

**Pull request merged**